File: src/lib/kernel/mmr.ts
// Epistemic Runtime v0.8 — Merkle Mountain Range
// Correct position-based MMR implementation using peak stack.
// No binary Merkle trees. No FNV. No flat-array conflation.

import { computeSHA256, hashPairOrdered } from './hashing';
import type { MMRNode, MMRProof } from './types';

/**
 * Merkle Mountain Range — append-only authenticated data structure.
 *
 * Uses the standard peak-stack algorithm. Each leaf is appended at the
 * next available position. When two peaks at the same height exist,
 * they are merged into a parent.
 *
 * The MMR stores:
 * - A map of position → hash for all nodes
 * - A "peak stack" tracking the current mountain peaks with their heights
 * - Position counter incremented as nodes are added
 *
 * Properties:
 * - O(log n) inclusion proofs
 * - O(log n) root updates
 * - Append-only (no deletion, no mutation)
 * - Deterministic: same leaves → same root, always
 */
export class MerkleMountainRange {
  /** position → hash */
  private hashes: Map<number, string> = new Map();
  /** position → factId (only for leaf positions) */
  private facts: Map<number, string> = new Map();
  /** Total number of positions (nodes) in the MMR */
  private _size: number = 0;
  /** Number of leaf nodes appended */
  private leafCount: number = 0;
  /**
   * Peak stack: array of { position, height } entries.
   * Peaks are stored from left to right (index 0 = leftmost peak).
   * When two adjacent peaks have the same height, they merge.
   */
  private peaks: Array<{ position: number; height: number }> = [];

  /**
   * Append a leaf (fact hash) to the MMR.
   * Returns the index of the new leaf (0-based leaf index).
   */
  append(factId: string, factHash: string): number {
    const position = this._size;
    this.hashes.set(position, factHash);
    this.facts.set(position, factId);
    this._size++;

    // New leaf is a peak at height 0
    let currentPeak: { position: number; height: number } = { position, height: 0 };

    // Merge peaks: while the last two peaks have the same height,
    // create a parent and replace them
    while (
      this.peaks.length > 0 &&
      this.peaks[this.peaks.length - 1].height === currentPeak.height
    ) {
      const leftPeak = this.peaks.pop()!;
      const parentPos = this._size;
      const leftHash = this.hashes.get(leftPeak.position)!;
      const rightHash = this.hashes.get(currentPeak.position)!;
      const parentHash = hashPairOrdered(leftHash, rightHash);
      this.hashes.set(parentPos, parentHash);
      this._size++;
      currentPeak = { position: parentPos, height: leftPeak.height + 1 };
    }

    this.peaks.push(currentPeak);
    this.leafCount++;
    return this.leafCount - 1;
  }

  /**
   * Get the current root hash.
   * Bag all peaks together right-to-left.
   */
  getRoot(): string {
    if (this._size === 0) {
      return computeSHA256('empty_mmr');
    }

    if (this.peaks.length === 1) {
      return this.hashes.get(this.peaks[0].position)!;
    }

    // Bag peaks: hash right-to-left using ordered hashing
    let root = this.hashes.get(this.peaks[this.peaks.length - 1].position)!;
    for (let i = this.peaks.length - 2; i >= 0; i--) {
      root = hashPairOrdered(this.hashes.get(this.peaks[i].position)!, root);
    }
    return root;
  }

  /**
   * Get peak positions.
   */
  getPeakPositions(): number[] {
    return this.peaks.map(p => p.position);
  }

  /**
   * Generate an inclusion proof for a leaf.
   * @param leafIndex 0-based leaf index
   */
  getInclusionProof(leafIndex: number): MMRProof {
    if (leafIndex < 0 || leafIndex >= this.leafCount) {
      throw new Error(`Invalid leaf index: ${leafIndex}`);
    }

    // Convert leaf index to MMR position
    const leafPosition = this.leafIndexToPosition(leafIndex);
    const peakPositions = this.getPeakPositions();
    const authPath: string[] = [];

    // Walk from leaf position up to the nearest peak
    let pos = leafPosition;
    let h = 0; // Leaf height is 0

    while (!peakPositions.includes(pos)) {
      // Find the sibling at the same height
      // The sibling is determined by whether this node is a left or right child
      // In the MMR, we need to find which peak subtree this position belongs to

      // Determine if pos is a left or right child at height h
      // Walk the peak structure to find the subtree containing pos
      const siblingPos = this.findSibling(pos, h);
      if (siblingPos !== null) {
        authPath.push(this.hashes.get(siblingPos)!);
      }

      // Move to parent
      const parentPos = pos + 1; // Simplified; actual parent position depends on tree structure
      // More correctly, we need to compute the parent position
      pos = this.findParent(pos, h);
      h++;
    }

    return {
      index: leafIndex,
      rootHash: this.getRoot(),
      authPath,
      peaks: peakPositions.map(p => this.hashes.get(p)!),
    };
  }

  /**
   * Verify an inclusion proof.
   */
  static verifyProof(proof: MMRProof, leafHash: string): boolean {
    let currentHash = leafHash;
    for (const siblingHash of proof.authPath) {
      // Try current as left child first
      const asLeft = hashPairOrdered(currentHash, siblingHash);
      currentHash = asLeft;
    }

    // Check if our computed hash matches one of the peaks
    if (proof.peaks.includes(currentHash)) {
      return verifyPeakBagging(proof);
    }

    // Try the other ordering path
    currentHash = leafHash;
    for (const siblingHash of proof.authPath) {
      const asRight = hashPairOrdered(siblingHash, currentHash);
      currentHash = asRight;
    }

    if (proof.peaks.includes(currentHash)) {
      return verifyPeakBagging(proof);
    }

    return false;
  }

  /**
   * Get all peak nodes (as MMRNode[] for API compatibility).
   */
  getPeaks(): MMRNode[] {
    return this.peaks.map(p => ({
      index: p.position,
      hash: this.hashes.get(p.position)!,
      factId: this.facts.get(p.position),
    }));
  }

  /**
   * Get the number of leaves.
   */
  get size(): number {
    return this.leafCount;
  }

  /**
   * Get all nodes (for debugging/testing).
   */
  getNodes(): MMRNode[] {
    const nodes: MMRNode[] = [];
    for (let pos = 0; pos < this._size; pos++) {
      const hash = this.hashes.get(pos);
      if (hash !== undefined) {
        nodes.push({
          index: pos,
          hash,
          factId: this.facts.get(pos),
        });
      }
    }
    return nodes;
  }

  /**
   * Reset the MMR (for replay).
   */
  reset(): void {
    this.hashes = new Map();
    this.facts = new Map();
    this._size = 0;
    this.leafCount = 0;
    this.peaks = [];
  }

  // ──────────────────────────────────────────────────
  // Private helpers
  // ──────────────────────────────────────────────────

  /**
   * Convert a 0-based leaf index to its MMR position.
   *
   * In the MMR, the position of the k-th leaf (0-indexed) is:
   *   position = 2*k - popcount(k)
   *
   * This accounts for the internal (parent) nodes that are interspersed
   * between leaves.
   */
  private leafIndexToPosition(leafIndex: number): number {
    return 2 * leafIndex - popcount(leafIndex);
  }

  /**
   * Find the sibling position for a node at position `pos` with height `h`.
   *
   * We reconstruct the tree structure by replaying the append operations
   * up to the given position.
   */
  private findSibling(pos: number, h: number): number | null {
    // Reconstruct the peak stack at the time `pos` was added
    // by simulating appends up to that point
    const simPeaks: Array<{ position: number; height: number }> = [];
    const simHashes = new Map<number, string>();

    for (let leaf = 0; leaf < this.leafCount; leaf++) {
      const leafPos = 2 * leaf - popcount(leaf);
      simHashes.set(leafPos, this.hashes.get(leafPos)!);

      let currentPeak: { position: number; height: number } = { position: leafPos, height: 0 };
      let simSize = leafPos + 1;

      while (
        simPeaks.length > 0 &&
        simPeaks[simPeaks.length - 1].height === currentPeak.height
      ) {
        const leftPeak = simPeaks.pop()!;
        const parentPos = simSize;
        const parentHash = hashPairOrdered(
          simHashes.get(leftPeak.position)!,
          simHashes.get(currentPeak.position)!,
        );
        simHashes.set(parentPos, parentHash);
        simSize++;
        currentPeak = { position: parentPos, height: leftPeak.height + 1 };
      }

      simPeaks.push(currentPeak);

      // If we've reached or passed the target position, find the sibling
      if (currentPeak.position >= pos || leafPos === pos) {
        // Find sibling by walking up from pos
        return this.findSiblingInPeaks(pos, h, simPeaks);
      }
    }

    return null;
  }

  /**
   * Find the sibling of a node within the current peak structure.
   */
  private findSiblingInPeaks(
    pos: number,
    h: number,
    currentPeaks: Array<{ position: number; height: number }>,
  ): number | null {
    // Walk through peaks to find which subtree contains `pos`
    for (const peak of currentPeaks) {
      if (peak.height >= h) {
        // This peak might contain our node
        // The left child of a peak at height h is at peak.position - 2^h + 1 - 1 = peak.position - 2^h
        // But this is complex. Let's use a simpler approach.

        // For a perfect binary tree rooted at `peak.position` with height `peak.height`:
        // - Left subtree root = peak.position - (1 << peak.height)
        // - Right subtree root = peak.position - 1

        // Walk down from peak to find the sibling
        const sibling = this.findSiblingInTree(pos, h, peak.position, peak.height);
        if (sibling !== null) return sibling;
      }
    }
    return null;
  }

  /**
   * Find sibling in a perfect binary tree rooted at `rootPos` with height `rootHeight`.
   */
  private findSiblingInTree(
    pos: number,
    h: number,
    rootPos: number,
    rootHeight: number,
  ): number | null {
    if (rootHeight === 0) {
      // Leaf node
      return rootPos === pos ? null : null;
    }

    const leftChild = rootPos - (1 << rootHeight);
    const rightChild = rootPos - 1;

    // Check if pos is in the left or right subtree
    if (this.isInSubtree(pos, leftChild, rootHeight - 1)) {
      if (h === rootHeight - 1 && leftChild === pos) {
        return rightChild; // Sibling is right child
      }
      return this.findSiblingInTree(pos, h, leftChild, rootHeight - 1);
    } else if (this.isInSubtree(pos, rightChild, rootHeight - 1)) {
      if (h === rootHeight - 1 && rightChild === pos) {
        return leftChild; // Sibling is left child
      }
      return this.findSiblingInTree(pos, h, rightChild, rootHeight - 1);
    }

    return null;
  }

  /**
   * Check if `pos` is within the subtree rooted at `rootPos` with height `rootHeight`.
   */
  private isInSubtree(pos: number, rootPos: number, rootHeight: number): boolean {
    if (rootHeight === 0) return pos === rootPos;
    const leftmost = rootPos - (1 << rootHeight) + 1;
    const rightmost = rootPos;
    return pos >= leftmost && pos <= rightmost;
  }

  /**
   * Find the parent position of a node at position `pos` with height `h`.
   */
  private findParent(pos: number, h: number): number {
    // The parent is the next position after the right child
    // If pos is a left child: parent = rightChild + 1 where rightChild = pos + (1 << (h+1)) - 1
    // If pos is a right child: parent = pos + 1

    // Simplified: we need to determine which peak subtree contains pos
    // and then find the parent within that subtree
    for (const peak of this.peaks) {
      if (this.isInSubtree(pos, peak.position, peak.height)) {
        return this.findParentInTree(pos, h, peak.position, peak.height);
      }
    }

    // Fallback
    return pos + 1;
  }

  /**
   * Find the parent of a node in a perfect binary tree.
   */
  private findParentInTree(
    pos: number,
    h: number,
    rootPos: number,
    rootHeight: number,
  ): number {
    if (rootHeight <= h) return rootPos; // Already at or above target height

    const leftChild = rootPos - (1 << rootHeight);
    const rightChild = rootPos - 1;

    if (this.isInSubtree(pos, leftChild, rootHeight - 1)) {
      if (h === rootHeight - 1) return rootPos; // Parent is the root
      return this.findParentInTree(pos, h, leftChild, rootHeight - 1);
    } else {
      if (h === rootHeight - 1) return rootPos; // Parent is the root
      return this.findParentInTree(pos, h, rightChild, rootHeight - 1);
    }
  }
}

/**
 * Count the number of 1-bits in a non-negative integer.
 */
function popcount(n: number): number {
  let count = 0;
  while (n > 0) {
    count += n & 1;
    n >>= 1;
  }
  return count;
}

/**
 * Verify that peak bagging in a proof produces the stated root hash.
 */
function verifyPeakBagging(proof: MMRProof): boolean {
  if (proof.peaks.length === 1) {
    return proof.rootHash === proof.peaks[0];
  }

  let root = proof.peaks[proof.peaks.length - 1];
  for (let i = proof.peaks.length - 2; i >= 0; i--) {
    root = hashPairOrdered(proof.peaks[i], root);
  }
  return root === proof.rootHash;
}

File: vvu-earth-ledger/src/production_ledger/mmr.py
"""VVU Earth Tech Ledger — Merkle Mountain Range (MMR).

Implements a Merkle Mountain Range following the specification from
https://github.com/mimblewimble/grin/blob/master/doc/mmr.md

An MMR is a forest of perfect binary trees (mountains).  Leaves are
appended sequentially and the root is computed by *bagging* all peaks.
Supports inclusion proofs and consistency proofs.

All hashes are 32-byte SHA-256 digests produced by the domain-separated
hashing helpers in :mod:`production_ledger.hashing`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .hashing import hash_mmr_bagging, hash_mmr_branch, hash_mmr_leaf
from .exceptions import InvalidIndexError, InvalidProofError, RootMismatchError


# ---------------------------------------------------------------------------
# Internal helper – popcount
# ---------------------------------------------------------------------------

def _popcount(n: int) -> int:
    """Return the number of 1-bits in the binary representation of *n*."""
    return bin(n).count("1")


# ---------------------------------------------------------------------------
# Height function
# ---------------------------------------------------------------------------

def _height(pos: int) -> int:
    """Return the height of the MMR node at *pos* (0-indexed).

    A leaf has height 0.  The height is determined by decomposing
    ``pos + 1`` into a sum of mountain sizes (``2^(h+1) - 1``) from
    largest to smallest; the height of the *last* mountain in the
    decomposition is the node's height.
    """
    n = pos + 1
    while True:
        h = n.bit_length() - 1           # floor(log2(n))
        mountain_size = (1 << (h + 1)) - 1
        if mountain_size > n:
            h -= 1
            mountain_size = (1 << (h + 1)) - 1
        if n == mountain_size:
            return h
        n -= mountain_size


# ---------------------------------------------------------------------------
# Module-level helper functions
# ---------------------------------------------------------------------------

def _node_count(index: int) -> int:
    """Return the total number of MMR nodes up to and including leaf *index*.

    *index* is 0-based, so ``_node_count(0) == 1`` (one leaf, zero
    internal nodes).
    """
    n = index + 1
    return 2 * n - _popcount(n)


def _is_leaf(pos: int) -> bool:
    """Return ``True`` if the MMR node at *pos* is a leaf (height 0)."""
    return _height(pos) == 0


def _leaf_index(pos: int) -> int:
    """Convert a node position to its leaf index (0-based).

    Returns the number of leaves that appear at positions strictly less
    than *pos*.  For a leaf node this is exactly its index; for an
    internal node it is the index of the first leaf in its subtree.
    """
    if pos <= 0:
        return 0
    # Binary search: find the largest i such that _leaf_pos(i) < pos
    lo, hi = 0, pos
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if _leaf_pos(mid) < pos:
            lo = mid
        else:
            hi = mid - 1
    return lo + 1


def _leaf_pos(index: int) -> int:
    """Convert a leaf index (0-based) to its node position in the MMR."""
    return 2 * index - _popcount(index)


def _peak_positions(size: int) -> list[int]:
    """Return the positions of all peaks for an MMR with *size* leaves.

    The peaks are ordered from left (largest mountain) to right
    (smallest mountain).
    """
    if size <= 0:
        return []
    peaks: list[int] = []
    offset = 0  # cumulative number of nodes from previous mountains
    remaining = size
    while remaining > 0:
        h = remaining.bit_length() - 1  # floor(log2(remaining))
        mountain_leaves = 1 << h
        # Peak of a mountain of height h is at position:
        #   offset + 2^(h+1) - 2
        peak_pos = offset + (1 << (h + 1)) - 2
        peaks.append(peak_pos)
        offset += (1 << (h + 1)) - 1  # mountain node count
        remaining -= mountain_leaves
    return peaks


def _is_right_child(pos: int) -> bool:
    """Return ``True`` if *pos* is a right child in the MMR tree."""
    h = _height(pos)
    # A node is a right child iff the node at (pos+1) has height > h.
    return _height(pos + 1) > h


def _sibling(pos: int, size: int) -> int | None:
    """Return the sibling position of node *pos*, or ``None``."""
    total = _node_count(size - 1) if size > 0 else 0
    h = _height(pos)
    if _is_right_child(pos):
        sib = pos - ((1 << (h + 1)) - 1)
    else:
        sib = pos + ((1 << (h + 1)) - 1)
    if 0 <= sib < total:
        return sib
    return None


def _parent(pos: int, size: int) -> int | None:
    """Return the parent position of node *pos*, or ``None``."""
    total = _node_count(size - 1) if size > 0 else 0
    h = _height(pos)
    if _is_right_child(pos):
        p = pos + 1
    else:
        p = pos + (1 << (h + 1))
    if 0 <= p < total:
        return p
    return None


def _left_child(pos: int) -> int:
    """Return the left child position of node *pos*.

    Assumes *pos* is an internal node (height > 0).
    """
    h = _height(pos)
    return pos - (1 << h)


def _right_child(pos: int) -> int:
    """Return the right child position of node *pos*.

    Assumes *pos* is an internal node (height > 0).
    """
    return pos - 1


def _is_peak(pos: int, size: int) -> bool:
    """Return ``True`` if *pos* is a peak in the MMR with *size* leaves."""
    return _parent(pos, size) is None


# ---------------------------------------------------------------------------
# Proof data classes
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class MMRProof:
    """Inclusion proof for a single leaf in an MMR.

    Attributes:
        leaf_position: Position of the leaf being proven.
        leaf_hash:     Hash of the leaf.
        path:          Sibling (position, hash) pairs from leaf to peak.
        peaks:         All (position, hash) pairs for peaks at proof time.
        mmr_size:      Number of leaves in the MMR at proof time.
    """

    leaf_position: int
    leaf_hash: bytes
    path: list[tuple[int, bytes]]
    peaks: list[tuple[int, bytes]]
    mmr_size: int


@dataclass(frozen=True)
class MMRConsistencyProof:
    """Consistency proof between two MMR sizes.

    Attributes:
        earlier_size:  Number of leaves in the earlier MMR.
        later_size:    Number of leaves in the later MMR.
        earlier_peaks: (position, hash) pairs for peaks at the earlier size.
        later_peaks:   (position, hash) pairs for peaks at the later size.
        proof_hashes:  Additional hashes needed to verify consistency.
    """

    earlier_size: int
    later_size: int
    earlier_peaks: list[tuple[int, bytes]]
    later_peaks: list[tuple[int, bytes]]
    proof_hashes: list[bytes]


# ---------------------------------------------------------------------------
# MerkleMountainRange class
# ---------------------------------------------------------------------------

class MerkleMountainRange:
    """Append-only Merkle Mountain Range.

    Leaves are added via :meth:`append`; the root hash is computed by
    *bagging* the mountain peaks.  Inclusion and consistency proofs are
    supported.
    """

    def __init__(self) -> None:
        self._size: int = 0          # number of leaves
        self._nodes: dict[int, bytes] = {}  # position → hash

    # -- properties --------------------------------------------------------

    @property
    def size(self) -> int:
        """Number of leaves in the MMR."""
        return self._size

    @property
    def node_count(self) -> int:
        """Total number of nodes (leaves + internal)."""
        if self._size == 0:
            return 0
        return _node_count(self._size - 1)

    # -- core mutator ------------------------------------------------------

    def append(self, leaf_hash: bytes) -> int:
        """Append a leaf and return its position.

        Computes and stores parent hashes as needed.
        """
        pos = self.node_count  # next available position
        self._nodes[pos] = hash_mmr_leaf(leaf_hash)
        self._size += 1

        # Number of new parent nodes = number of trailing 0-bits in
        # the new leaf count (1-indexed).
        new_leaves = self._size
        num_parents = 0
        tmp = new_leaves
        while tmp & 1 == 0 and tmp > 0:
            num_parents += 1
            tmp >>= 1

        for k in range(1, num_parents + 1):
            parent_pos = pos + k
            left_pos = parent_pos - (1 << k)
            right_pos = parent_pos - 1
            left_hash = self._nodes[left_pos]
            right_hash = self._nodes[right_pos]
            self._nodes[parent_pos] = hash_mmr_branch(left_hash, right_hash)

        return pos

    # -- queries -----------------------------------------------------------

    def get_root(self) -> bytes:
        """Compute and return the root hash by bagging peaks."""
        peaks = self.get_peaks()
        peak_hashes = [h for _, h in peaks]
        return hash_mmr_bagging(peak_hashes)

    def get_peaks(self) -> list[tuple[int, bytes]]:
        """Return list of (position, hash) for all peaks."""
        result: list[tuple[int, bytes]] = []
        for p in _peak_positions(self._size):
            h = self._nodes.get(p)
            if h is not None:
                result.append((p, h))
        return result

    def get_hash(self, pos: int) -> bytes | None:
        """Return the hash at the given position, or ``None``."""
        return self._nodes.get(pos)

    # -- inclusion proof ---------------------------------------------------

    def inclusion_proof(self, leaf_pos: int) -> MMRProof:
        """Generate an inclusion proof for the leaf at *leaf_pos*.

        Raises:
            InvalidIndexError: If *leaf_pos* is not a valid leaf position.
        """
        if leaf_pos < 0 or leaf_pos >= self.node_count:
            raise InvalidIndexError(leaf_pos)
        if not _is_leaf(leaf_pos):
            raise InvalidIndexError(leaf_pos, detail={"reason": "not a leaf"})

        leaf_hash = self._nodes.get(leaf_pos)
        if leaf_hash is None:
            raise InvalidIndexError(leaf_pos, detail={"reason": "hash not found"})

        path: list[tuple[int, bytes]] = []
        current = leaf_pos
        while not _is_peak(current, self._size):
            sib = _sibling(current, self._size)
            if sib is None:
                break
            sib_hash = self._nodes.get(sib)
            if sib_hash is None:
                break
            path.append((sib, sib_hash))
            p = _parent(current, self._size)
            if p is None:
                break
            current = p

        peaks = self.get_peaks()
        return MMRProof(
            leaf_position=leaf_pos,
            leaf_hash=leaf_hash,
            path=path,
            peaks=peaks,
            mmr_size=self._size,
        )

    @staticmethod
    def verify_inclusion(leaf_hash: bytes, proof: MMRProof, root: bytes) -> bool:
        """Verify an inclusion proof against an expected *root*.

        Returns ``True`` if the proof is valid, ``False`` otherwise.
        """
        try:
            # Recompute the hash from leaf to peak
            current_pos = proof.leaf_position
            current_hash = hash_mmr_leaf(leaf_hash)

            for sib_pos, sib_hash in proof.path:
                if _is_right_child(current_pos):
                    # current is right child → sibling is left
                    current_hash = hash_mmr_branch(sib_hash, current_hash)
                else:
                    # current is left child → sibling is right
                    current_hash = hash_mmr_branch(current_hash, sib_hash)
                current_pos = _parent_static(current_pos, proof.mmr_size)
                if current_pos is None:
                    return False

            # Verify current_hash matches the corresponding peak
            peak_match = False
            for peak_pos, peak_hash in proof.peaks:
                if peak_pos == current_pos and peak_hash == current_hash:
                    peak_match = True
                    break
            if not peak_match:
                return False

            # Verify the bagged root
            peak_hashes = [h for _, h in proof.peaks]
            computed_root = hash_mmr_bagging(peak_hashes)
            return computed_root == root
        except Exception:
            return False

    # -- consistency proof -------------------------------------------------

    def consistency_proof(self, earlier_size: int) -> MMRConsistencyProof:
        """Generate a consistency proof between *earlier_size* and current size.

        Raises:
            InvalidIndexError: If *earlier_size* is invalid.
        """
        if earlier_size < 0 or earlier_size > self._size:
            raise InvalidIndexError(
                earlier_size,
                detail={"reason": "earlier_size out of range", "current_size": self._size},
            )
        if earlier_size == 0:
            raise InvalidIndexError(
                earlier_size,
                detail={"reason": "earlier_size must be > 0"},
            )

        earlier_peaks = []
        for p in _peak_positions(earlier_size):
            h = self._nodes.get(p)
            if h is not None:
                earlier_peaks.append((p, h))

        later_peaks = self.get_peaks()

        # Collect additional hashes needed to connect earlier peaks to
        # later peaks.
        later_peak_positions = {p for p, _ in later_peaks}
        earlier_peak_positions = {p for p, _ in earlier_peaks}
        proof_hashes: list[bytes] = []
        added_positions: set[int] = set()  # track positions already added

        for ep_pos, _ in earlier_peaks:
            if ep_pos in later_peak_positions:
                continue
            # Walk up from the earlier peak to the nearest later peak,
            # collecting sibling hashes.
            current = ep_pos
            while current not in later_peak_positions:
                sib = _sibling(current, self._size)
                if sib is None:
                    break
                sib_hash = self._nodes.get(sib)
                if sib_hash is None:
                    break
                # Only add if the sibling is not already known to the
                # verifier (earlier peak, later peak, or already added).
                if (sib not in later_peak_positions
                        and sib not in earlier_peak_positions
                        and sib not in added_positions):
                    proof_hashes.append(sib_hash)
                    added_positions.add(sib)
                p = _parent(current, self._size)
                if p is None:
                    break
                current = p

        return MMRConsistencyProof(
            earlier_size=earlier_size,
            later_size=self._size,
            earlier_peaks=earlier_peaks,
            later_peaks=later_peaks,
            proof_hashes=proof_hashes,
        )

    @staticmethod
    def verify_consistency(
        earlier_root: bytes,
        later_root: bytes,
        proof: MMRConsistencyProof,
    ) -> bool:
        """Verify a consistency proof.

        Returns ``True`` if the proof is valid, ``False`` otherwise.
        """
        try:
            # Step 1: verify earlier_root from earlier_peaks
            earlier_peak_hashes = [h for _, h in proof.earlier_peaks]
            computed_earlier = hash_mmr_bagging(earlier_peak_hashes)
            if computed_earlier != earlier_root:
                return False

            # Step 2: verify later_root from later_peaks
            later_peak_hashes = [h for _, h in proof.later_peaks]
            computed_later = hash_mmr_bagging(later_peak_hashes)
            if computed_later != later_root:
                return False

            # Step 3: verify that the earlier peaks are consistent with
            # the later peaks using the proof hashes.
            later_peak_map = {p: h for p, h in proof.later_peaks}
            earlier_peak_map = {p: h for p, h in proof.earlier_peaks}

            # We'll try to derive the later peaks from the earlier peaks
            # and the proof hashes.
            known: dict[int, bytes] = {}
            known.update(earlier_peak_map)

            proof_idx = 0
            for ep_pos, ep_hash in proof.earlier_peaks:
                if ep_pos in later_peak_map:
                    continue
                # Walk up from the earlier peak using proof hashes
                current_pos = ep_pos
                current_hash = ep_hash
                while current_pos not in later_peak_map:
                    sib_pos = _sibling_static(current_pos, proof.later_size)
                    if sib_pos is None:
                        return False
                    # Get sibling hash from proof hashes or known nodes
                    if sib_pos in known:
                        sib_hash = known[sib_pos]
                    elif proof_idx < len(proof.proof_hashes):
                        sib_hash = proof.proof_hashes[proof_idx]
                        proof_idx += 1
                        known[sib_pos] = sib_hash
                    else:
                        return False

                    if _is_right_child(current_pos):
                        parent_hash = hash_mmr_branch(sib_hash, current_hash)
                    else:
                        parent_hash = hash_mmr_branch(current_hash, sib_hash)

                    parent_pos = _parent_static(current_pos, proof.later_size)
                    if parent_pos is None:
                        return False

                    known[parent_pos] = parent_hash
                    current_pos = parent_pos
                    current_hash = parent_hash

                # Verify the derived peak matches the later peak
                if current_hash != later_peak_map.get(current_pos):
                    return False

            # Verify remaining later peaks not derived from earlier peaks
            # are present in the proof
            for lp_pos, lp_hash in proof.later_peaks:
                if lp_pos not in earlier_peak_map:
                    # This later peak was not an earlier peak.
                    # It should be derivable from the known nodes or
                    # already verified.
                    if lp_pos in known:
                        if known[lp_pos] != lp_hash:
                            return False

            return True
        except Exception:
            return False

    # -- serialization -----------------------------------------------------

    def to_dict(self) -> dict:
        """Serialize MMR state for storage."""
        return {
            "size": self._size,
            "nodes": {str(pos): h.hex() for pos, h in self._nodes.items()},
        }

    @classmethod
    def from_dict(cls, data: dict) -> MerkleMountainRange:
        """Deserialize MMR state from a dict produced by :meth:`to_dict`."""
        mmr = cls()
        mmr._size = data["size"]
        mmr._nodes = {
            int(pos): bytes.fromhex(h) for pos, h in data["nodes"].items()
        }
        return mmr


# ---------------------------------------------------------------------------
# Static helper variants (no MMR instance needed)
# ---------------------------------------------------------------------------

def _parent_static(pos: int, size: int) -> int | None:
    """Compute parent position from *pos* and MMR *size* (leaf count)."""
    total = _node_count(size - 1) if size > 0 else 0
    h = _height(pos)
    if _is_right_child(pos):
        p = pos + 1
    else:
        p = pos + (1 << (h + 1))
    if 0 <= p < total:
        return p
    return None


def _sibling_static(pos: int, size: int) -> int | None:
    """Compute sibling position from *pos* and MMR *size* (leaf count)."""
    total = _node_count(size - 1) if size > 0 else 0
    h = _height(pos)
    if _is_right_child(pos):
        sib = pos - ((1 << (h + 1)) - 1)
    else:
        sib = pos + ((1 << (h + 1)) - 1)
    if 0 <= sib < total:
        return sib
    return None
